package building infrastructure

hairmare opened the pull request:
I did some packaging work and am putting it here to get some early review. I'll de-WIP once it builds properly on CentOS 7.

* [x] build rpm with code and entrypoint
* [x] create dynamic user/group during package installation
* [x] move package to suisa_sendemeldung to make using setuptools possible
* [x] systemd service and timer for running this once a month
* [x] figure out how to template specfile (versioning) in a way that is runnable on OBS
* [x] add docs
* [x] switch to proper x.y.z versions?
* [x] add requirements.txt
* [x] install default sysconfig
* [x] install default config file
* [x] remove clunky parts from sysconfig
* [x] figure out a sensible way to proceed about missing python34-configargparse (python2-configargparse does exist though)

am5z commented:
Thank you, looks great so far! Now we should think about implementing a configuration file. That's probably what I'll do next...

hairmare commented:
@am5z With this PR it's also possible to specify `OPTIONS` in `/etc/sysconfig/suisa_sendemeldung` like so:

```
OPTIONS="--my --awesome --options"
```

Another option would be to make everything configurable in env variables according to the [Twelve-Factor App Methodology](https://12factor.net/).

hairmare commented:
I will be using the following `_service` on OBS (This can't be managed through GitHub).

```xml
<services>
  <service name="tar_scm">
    <param name="url">git://github.com/radiorabe/suisa_sendemeldung.git</param>
    <param name="scm">git</param>
    <param name="revision">feature/packaging</param>
    <param name="versionformat">@PARENT_TAG@.%h</param>
  </service>
  <service name="extract_file">
    <param name="archive">*.tar</param>
    <param name="files">*/suisa_sendemeldung.spec</param>
  </service>
  <service name="download_files" />
  <service name="recompress">
    <param name="file">*.tar</param>
    <param name="compression">gz</param>
  </service>
  <service name="set_version"/>
</services>
```

I will also create a stand-alone OBS Project/Repo with a Project Config containing this: 
`Release: 0.<CI_CNT>.<B_CNT>.rabe` so that the Release part of the version zero prefixed.

This will make package versions based on the last reachable tag going up the git dag and the actual short commit hash at the end so we are always sure what went into a package.

I'll configure the GitHub OBS integration so it only triggers builds when new tags get pushed. This way we will only be building each tag once.

With this package versions will look like so:

| Tag | Full Commit | Package Version |
| ---- | ---- | ---- |
| 1.0.0 | cafefb04df6d5b6d421ed652036110e830dbf3b8 | 1.0.0.cafefb0-0.<CI_CNT>.<B_CNT>.rabe |

Where CI_CNT is a number of commits and B_CNT is a number of rebuilds.

hairmare commented:
I added an example `/etc/sysconfig/suisa_sendemeldung` but it looks quite clunky (sysconfig are usually used for simpler stuff like `--no-daemonize`) so I'll probably strip it down to a bare `OPTIONS=""` if you decide to implement a real configuration file.

am5z commented:
Hey @hairmare, sorry for letting you wait so long. I couldn't invest much time lately into this project.

I quite like the idea of the Twelve-Factor App to store configuration in the environment or at least make it possible to configure the script that way. I'm thinking about adding [ConfigArgParse](https://github.com/bw2/ConfigArgParse) to the project. It's a drop-in replacement for `argparse` that supports both environment variables and a configuration file with proper precedence. What do you think?

Is it possible and ideal to store environment variables in `/etc/sysconfig/suisa_sendemeldung`? If not, where else would we put the environment variables?

hairmare commented:
ConfigArgParse looks like a good solution. Putting the env vars in `/etc/sysconfig/suisa_sendemeldung` is fine IMHO. The main reason I considered the conf file in this PR to be clunky was due to the stringing together of the longish args string here:

https://github.com/radiorabe/suisa_sendemeldung/blob/7be62a430df40e97707bd11d4e26b79a7eee9e9f/etc/sysconfig/suisa_sendemeldung#L13-L14

If we move all of that to proper env vars everything will look much better.

am5z commented:
I pushed the changes needed for configuration files and environment variables to #7. @hairmare what do you think?

Regarding docs: I am currently expanding `README.md`. Can we move this point out to a separat PR?

hairmare commented:
> Regarding docs: I am currently expanding README.md. Can we move this point out to a separat PR?

Sure, the task in the issue is a reminder for me that I document the changes that this PR introduces (ie. adding some simple setup instructions for CentOS users, ...).

hairmare commented:
I fixed the [Anitya entry for ConfigArgParse](https://release-monitoring.org/project/14814/) by adding Fedora and Fedora EPEL to the Distro Mappings which lead to a [package bump bug](https://bugzilla.redhat.com/show_bug.cgi?id=1643700) in Bugzilla.
I'll ask @fabaff (the Fedora Maintainer of configargparse) if he would be willing to release it into EPEL once the bump is through. [This PR](https://github.com/bw2/ConfigArgParse/pull/117) indicates that it should work fine in python34 from EPEL.

am5z commented:
Heya, I added what I've got so far in terms of documentation in #9. @hairmare, feel free to add your comments.

hairmare commented on etc/systemd/suisa_sendemeldung.timer:
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Run ACRCloud client for SUISA reporting on first monday of every month
```
s/mond/Mond/

hairmare commented:
I rebased this, added some docs and, built `python34-configargparse-0.12.0-0.7.1.rabe.noarch.rpm` on OBS to take care of the remaining issues still open.

hairmare commented:
I de-WIPed it. After we merge this we should tag a release (ie. 0.1.0) through GitHub right away (by tagging the merge commit). I'll then finalize configuring the prepared OBS package and we should have installable RPMs pretty soon afterwards.

am5z commented:
Looks good! The only issue I see is that because `suisa_sendemeldung.conf` has been moved the link in `README.md` will become broken. Could you fix that?

am5z commented:
Great! Thank you very much!
